src/main/groovy/liquibase/harness/data/ChangeDataTest.groovy

```diff
@@ -0,0 +1,97 @@
+        String checkingSql = cleanSql(getExpectedSqlFileContent(testInput.changeData,
+                testInput.databaseName, testInput.version, "liquibase/harness/data/checkingSql"))
+        String expectedResultSet = getExpectedJSONFileContent(testInput.changeData, testInput.databaseName,
+                testInput.version, "liquibase/harness/data/expectedResultSet")
+
+        and: "skip testcase if it's invalid for this combination of db type and/or version"
+        Assume.assumeTrue(expectedSql, expectedSql == null || !expectedSql.toLowerCase().contains("invalid test"))
+
+        and: "fail test if expectedResultSet is not provided"
+        assert expectedResultSet != null: "No expectedResultSet for ${testInput.changeData} against " +
+                "${testInput.database.shortName} " +
+                "${testInput.database.databaseMajorVersion}.${testInput.database.databaseMinorVersion}"
+
+        and: "fail test if checkingSql is not provided"
+        assert checkingSql != null: "No checkingSql for ${testInput.changeData} against " +
+                "${testInput.database.shortName} " +
+                "${testInput.database.databaseMajorVersion}.${testInput.database.databaseMinorVersion}"
+
+        when: "get sql that is generated for changeset"
+        def generatedSql = cleanSql(toSqlFromLiquibaseChangeSets(liquibase))
+
+        then: "verify expected sql matches generated sql"
+        if (expectedSql != null && !testInput.pathToChangeLogFile.endsWith(".sql")) {
+            assert generatedSql == expectedSql: "Expected SQL does not match actual sql. " +
+                    "Deleting the existing expectedSql file will test that the new SQL works correctly " +
+                    "and will auto-generate a new version if it passes"
+            if (!TestConfig.instance.revalidateSql) {
+                return //sql is right. Nothing more to test
+            }
+        }
+
+        def connection = testInput.database.getConnection()
+
+        assert connection instanceof JdbcConnection: "We cannot verify the following SQL works " +
+                "because the database is offline:\n${generatedSql}"
+
+        when: "apply changeSet to DB,"
+        try {
+            liquibase.update(testInput.context)
+        } catch (Throwable throwable) {
+            println "Error executing sql. If this is expected to be invalid SQL for this database/version, " +
+                    "create an 'expectedSql/${testInput.database.shortName}/${testInput.changeData}.sql' file that starts with " +
+                    "'INVALID TEST' and an explanation of why."
+            throwable.printStackTrace()
+            Assert.fail throwable.message
+        }
+
+        then: "obtain resultSet form the statement, compare expected resultSet to generated resultSet, apply rollback"
+        try {
+            def statement = connection.createStatement()
```

Comment:
nit
redundant variables. If they are used ones, inline subsequent calls. Check refactoring techniques for references

Reply:
++